File: project/core/lib/common/src/main/java/vdi/core/health/RemoteDependencies.java
package vdi.core.health;

import java.util.*;
import java.util.function.Supplier;

public final class RemoteDependencies {
  private static final Deque<Dependency> REGISTRY = new ArrayDeque<>();

  public static void register(Dependency dependency) {
    REGISTRY.addLast(dependency);
  }

  public static void register(
    String name,
    String host,
    int port,
    String protocol,
    Supplier<Map<String, Object>> extraFieldSupplier
  ) {
    if (extraFieldSupplier == null)
      REGISTRY.addLast(new StaticDependency(name, protocol, host, port, Collections.emptyMap()));
    else
      REGISTRY.addLast(new DynamicDependency(name, protocol, host, port, extraFieldSupplier));
  }

  public static Iterator<Dependency> iterator() {
    return new Iterator<>() {
      private final Iterator<Dependency> raw = REGISTRY.iterator();

      @Override
      public boolean hasNext() {
        return raw.hasNext();
      }

      @Override
      public Dependency next() {
        return raw.next();
      }
    };
  }
}

File: lib/health/src/main/kotlin/vdi/health/RemoteDependencies.kt
package vdi.health

import io.klibs.collections.Deque

object RemoteDependencies : Iterable<Dependency> {
  private val registry = Deque<Dependency>()

  fun register(dependency: Dependency) {
    registry.append(dependency)
  }

  fun register(
    name: String,
    host: String,
    port: UShort,
    protocol: String = "",
    extraProvider: (() -> Map<String, Any>)? = null,
  ) {
    if (extraProvider != null)
      registry.append(DynamicDependency(name, protocol, host, port, extraProvider))
    else
      registry.append(StaticDependency(name, protocol, host, port))
  }

  override fun iterator(): Iterator<Dependency> {
    return object : Iterator<Dependency> {
      private val raw = registry.iterator()
      override fun hasNext() = raw.hasNext()
      override fun next() = raw.next()
    }
  }
}
